### PR TITLE
save and load pulled containers state

### PR DESCRIPTION
**Pull request by sharanyad**
### Summary
This PR is a continuation of https://github.com/aws/amazon-ecs-agent/pull/2716 . So please review only the 2nd commit. 

- saves and loads the new pulled containers state to and from agent's boltdb state. 
- It also removes it from state when container moves to `CREATED` state. The removal is also attempted when task is stopped, in case containers didn't move to `CREATED` before moving to `STOPPED` state.
- updates the already added pulled container state when known status of container is updated to `PULLED`

### Implementation details
- Add a new bucket named `pulledcontainers` in boltdb state 
- Introduce methods `SavePulledContainer`, `DeletePulledContainer`, `GetPulledContainers` in data client for saving, removing and loading data from boltdb
- update the already added pulled container state when known status of container is updated to `PULLED` in `managedTask.handleContainerChange` method.
- today pulled containers are added to containers table in agent state here https://github.com/aws/amazon-ecs-agent/blob/1489adfaec05c559ea4e1490774d1f86261cf936/agent/engine/task_manager.go#L422 As a result, when agent restarts, the pulled container is added to non-pulled containers state, which then is attempted to obtain network metadata here https://github.com/aws/amazon-ecs-agent/blob/84321fcb15787f1d3a3f8aed80002d6d7d080ec6/agent/handlers/v4/container_metadata_handler.go#L74 and fails here https://github.com/aws/amazon-ecs-agent/blob/84321fcb15787f1d3a3f8aed80002d6d7d080ec6/agent/handlers/v4/container_metadata_handler.go#L85 
Fixed that in this PR, by skipping if container's known states is `PULLED` and container ID is empty. 

### Testing

New tests cover the changes: 
- Unit tests yet to be updated 
- Manual testing done: 
TMDE V4 response:
```
curl ${ECS_CONTAINER_METADATA_URI_V4}/task
{"Cluster":"pull-state","TaskARN":"arn:aws:ecs:us-west-2:074626758757:task/pull-state/2f355c277e1d4bbdb6cfe8da557b98fb","Family":"dep-check","Revision":"1","DesiredStatus":"RUNNING","KnownStatus":"NONE","PullStartedAt":"2020-11-11T14:56:03.500537906Z","PullStoppedAt":"2020-11-11T14:56:04.87932092Z","AvailabilityZone":"us-west-2a","LaunchType":"EC2","Containers":[{"DockerId":"5e733312cebf2b606e9a41af031f42a596d5a025ae34a407d359c9a6d95b7f30","Name":"network","DockerName":"ecs-dep-check-1-network-daebabb79381aea78901","Image":"ubuntu","ImageID":"sha256:d70eaf7277eada08fca944de400e7e4dd97b1262c06ed2b1011500caa4decaf1","Labels":{"com.amazonaws.ecs.cluster":"pull-state","com.amazonaws.ecs.container-name":"network","com.amazonaws.ecs.task-arn":"arn:aws:ecs:us-west-2:074626758757:task/pull-state/2f355c277e1d4bbdb6cfe8da557b98fb","com.amazonaws.ecs.task-definition-family":"dep-check","com.amazonaws.ecs.task-definition-version":"1"},"DesiredStatus":"RUNNING","KnownStatus":"RUNNING","Limits":{"CPU":0,"Memory":512},"CreatedAt":"2020-11-11T14:56:04.777143666Z","StartedAt":"2020-11-11T14:56:05.421289274Z","Type":"NORMAL","ContainerARN":"arn:aws:ecs:us-west-2:074626758757:container/fb246e8e-4dab-4c3a-b7f6-f753c1ee1b30","Networks":[{"NetworkMode":"bridge","IPv4Addresses":["172.17.0.2"]}]},{"DockerId":"","Name":"customer","DockerName":"","Image":"amazonlinux:latest","ImageID":"sha256:ba2cc467a2bce7ffc9e2d6e38a87f6ec89c1c93bc83ff575e8021bb592f9ad2e","DesiredStatus":"RUNNING","KnownStatus":"PULLED","Limits":{"CPU":0,"Memory":512},"Type":"NORMAL","ContainerARN":"arn:aws:ecs:us-west-2:074626758757:container/7a400635-6cd2-4158-974d-bfab9885abab"}]} 
``` 
Restart agent now 
TMDE response after agent restart (pulled container is still seen)
```
{"Cluster":"pull-state","TaskARN":"arn:aws:ecs:us-west-2:074626758757:task/pull-state/2f355c277e1d4bbdb6cfe8da557b98fb","Family":"dep-check","Revision":"1","DesiredStatus":"RUNNING","KnownStatus":"NONE","PullStartedAt":"2020-11-11T14:56:03.500537906Z","PullStoppedAt":"2020-11-11T14:56:04.87932092Z","AvailabilityZone":"us-west-2a","LaunchType":"EC2","Containers":[{"DockerId":"5e733312cebf2b606e9a41af031f42a596d5a025ae34a407d359c9a6d95b7f30","Name":"network","DockerName":"ecs-dep-check-1-network-daebabb79381aea78901","Image":"ubuntu","ImageID":"sha256:d70eaf7277eada08fca944de400e7e4dd97b1262c06ed2b1011500caa4decaf1","Labels":{"com.amazonaws.ecs.cluster":"pull-state","com.amazonaws.ecs.container-name":"network","com.amazonaws.ecs.task-arn":"arn:aws:ecs:us-west-2:074626758757:task/pull-state/2f355c277e1d4bbdb6cfe8da557b98fb","com.amazonaws.ecs.task-definition-family":"dep-check","com.amazonaws.ecs.task-definition-version":"1"},"DesiredStatus":"RUNNING","KnownStatus":"RUNNING","Limits":{"CPU":0,"Memory":512},"CreatedAt":"2020-11-11T14:56:04.777143666Z","StartedAt":"2020-11-11T14:56:05.421289274Z","Type":"NORMAL","ContainerARN":"arn:aws:ecs:us-west-2:074626758757:container/fb246e8e-4dab-4c3a-b7f6-f753c1ee1b30","Networks":[{"NetworkMode":"bridge","IPv4Addresses":["172.17.0.2"]}]},{"DockerId":"","Name":"customer","DockerName":"","Image":"amazonlinux:latest","ImageID":"sha256:ba2cc467a2bce7ffc9e2d6e38a87f6ec89c1c93bc83ff575e8021bb592f9ad2e","DesiredStatus":"RUNNING","KnownStatus":"PULLED","Limits":{"CPU":0,"Memory":512},"Type":"NORMAL","ContainerARN":"arn:aws:ecs:us-west-2:074626758757:container/7a400635-6cd2-4158-974d-bfab9885abab"}]}
```
example of `agent.db` with pulled container added:
```
| Path: pulledcontainers → ea86e60a718745eb850fd6afefa4eb1f-customer
| Key: ea86e60a718745eb850fd6afefa4eb1f-customer
| Value:
| {
|   "Container": {
|     "ApplyingError": null,
|     "Command": [
|       "sleep 100"
|     ],
|     "ContainerArn": "arn:aws:ecs:us-west-2:074626758757:contai....
….
….
|     "KnownStatus": "PULLED",
```
once container moves to created:
```
| Path: pulledcontainers
| Buckets: 0
| Pairs: 0
```

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/docker_task_engine.go [hunk 1]
```diff
@@ -873,6 +873,12 @@ func (engine *DockerTaskEngine) pullAndUpdateContainerReference(task *apitask.Ta
 		return metadata
 	}
 	pullSucceeded := metadata.Error == nil
+	if pullSucceeded && engine.cfg.ContainerPullInParallel.Enabled() {
```
Comment (ubhattacharjya) [inline agent/engine/docker_task_engine.go:876]: I think the config var check is not needed now as now all pulled containers will be added to state

Comment (sharanyad) [reply agent/engine/docker_task_engine.go:876]: this is from 1st commit in the PR, please review only the 2nd commit (details in PR description)

Comment (ubhattacharjya) [reply agent/engine/docker_task_engine.go:876]: oh yes...thanks

File: agent/engine/data.go [hunk 2]
```diff
@@ -220,6 +243,21 @@ func (engine *DockerTaskEngine) removeTaskData(task *apitask.Task) {
 			seelog.Errorf("Failed to remove data for container %s: %v", c.Name, err)
 		}
 	}
+	containers, ok := engine.state.PulledContainerMapByArn(task.Arn)
+	if ok {
```
Comment (angelcar) [inline agent/engine/data.go:247]: nit: not necessary to check for `ok==true`, you can just iterate over containers, even if it's nil it won't cause errors

File: agent/data/client.go [hunk 3]
```diff
@@ -43,6 +44,7 @@ var (
 	buckets = []string{
 		imagesBucketName,
 		containersBucketName,
+		pulledContainersBucketName,
```
Comment (fenxiong) [inline agent/data/client.go:47]: is a separate bucket for pulled container necessary? each container is already added to the `containers` bucket when it transitions to pulled: https://github.com/aws/amazon-ecs-agent/blob/478c4ec00a43b4713a0acc811cfd1e2d70307f63/agent/engine/task_manager.go#L422. saving it to another bucket means adding an extra save for every container while it doesn't save anything new. To ensure that pulled containers can be correctly loaded into state when agent restarts, it seems to me that you don't need anything change for container saving/removing, while for loading, you can make a change here https://github.com/aws/amazon-ecs-agent/blob/478c4ec00a43b4713a0acc811cfd1e2d70307f63/agent/engine/data.go#L68-L77 to check the container status, and add pulled container to state with `state.AddPulledContainer` while adding other containers with `state.AddContainer`

**Comment by sharanyad** [comment]
closing in favor of https://github.com/aws/amazon-ecs-agent/pull/2718

**Pull request closed without merging**